napari/components/_dims/dims.py

```diff
@@ -0,0 +1,294 @@
+from napari.util.event import EmitterGroup
+
+
+class DimsMode(Enum):
+    Point = 0
+    Interval = 1
+
+
+class Dims(Component):
+
+    def __init__(self, init_max_dims=0):
+        """
+        Dimensions object modelling multi-dimensional slicing, cropping, and displaying in Napari
+        Parameters
+        ----------
+        init_max_dims : initial number of dimensions
+        """
+        super().__init__()
+
+        # Events:
+        self.changed = EmitterGroup(source=self,
+                                    auto_connect=True,
+                                    axis=None,
+                                    nbdims=None)
+
+        self.range = []
+        self.point = []
+        self.interval = []
+        self.mode = []
+        self.display = []
+
+        self._ensure_axis_present(init_max_dims - 1)
+
+    def __str__(self):
+        return "~~".join([str(self.range),
+                         str(self.point),
+                         str(self.interval),
+                         str(self.mode),
+                         str(self.display)])
+
+    @property
+    def num_dimensions(self):
+        """
+        Returns the number of dimensions
+        Returns numebr of dimensions
+        -------
+
+        """
+        return len(self.point)
+
+    @num_dimensions.setter
+    def num_dimensions(self, num_dimensions):
+        if self.num_dimensions < num_dimensions:
+            self._ensure_axis_present(num_dimensions - 1)
+        elif self.num_dimensions > num_dimensions:
+            self._trim_nb_dimensions(num_dimensions)
+
+    def set_all_ranges(self, all_ranges: Sequence[Union[int, float]]):
+        """
+        Sets ranges for all dimensions
+        Parameters
+        ----------
+        ranges : tuple or
+        """
+        nb_dim = len(all_ranges)
+        modified_dims = self._ensure_axis_present(nb_dim-1, no_event=True)
+        self.range=all_ranges
+
+        self.changed.nbdims()
+        for axis_changed in modified_dims:
+            self.changed.axis(axis=axis_changed)
+
+    def set_range(self, axis: int, range: Sequence[Union[int, float]]):
+        """
+        Sets the range (min, max, step) for a given axis (dimension)
+        Parameters
+        ----------
+        axis : dimension index
+        range : (min, max, step) tuple
+        """
+        self._ensure_axis_present(axis)
+        if self.range[axis] != range:
+            self.range[axis] = range
+            self.changed.axis(axis=axis)
+
+    def get_range(self, axis: int):
+        """
+        Returns the point at which this dimension is sliced
+        Parameters
+        ----------
+        axis : dimension index
+        """
+        return self.range[axis]
+
+    def set_point(self, axis: int, value: Union[int, float]):
+        """
+        Sets thepoint at which to slice this dimension
+        Parameters
+        ----------
+        axis : dimension index
+        value :
```

Comment:
I'm just gonna stop making docstring comments here. You should take a look and fix all of the according to the [numpydoc specification](https://numpydoc.readthedocs.io/en/latest/format.html) or I can do it for you.

Reply:
gone through and done this as best i could